### DISCUSSION TITLE
Reading Characteristic Notifies

### DISCUSSION THREAD

**Discussion by htcrane** [hbldh/bleak#965] [Q&A]
I was playing around with the async_callback_with_queue example to test out some larger data transfers (512 bytes). My peripheral device has a characteristic with notify and read attributes. When the peripheral updates the data, a notification is sent. I use this as an indication to go make a characteristic read to retrieve all 512 bytes. I do this by reading within the notification handler. However, when I do, the read causes the notify to trigger again and again. 

I've tried looking through forms and I don't believe a characteristic read should trigger a notification on the central running the python script. Is this behavior expected?

```
import sys
import time
import platform
import asyncio
import logging

from bleak import BleakClient

logger = logging.getLogger(__name__)

ADDRESS = (
    "24:71:89:cc:09:05"
    if platform.system() != "Darwin"
    else "72E269E5-DE4E-21FB-C792-F2B54DCF1DC1"
)
CHARACTERISTIC_UUID = "7eb661a9-6fbc-4b83-9acc-c58e165f3ba1"


async def run_ble_client(address: str, char_uuid: str, queue: asyncio.Queue):
    async def callback_handler(sender, data):
        await queue.put((time.time(), data))
        item = await client.read_gatt_char(char_uuid)
        print("Read:", item)

    async with BleakClient(address) as client:
        logger.info(f"Connected: {client.is_connected}")
        await client.start_notify(char_uuid, callback_handler)
        await asyncio.sleep(20.0)
        await client.stop_notify(char_uuid)
        # Send an "exit command to the consumer"
        await queue.put((time.time(), None))


async def run_queue_consumer(queue: asyncio.Queue):
    while True:
        # Use await asyncio.wait_for(queue.get(), timeout=1.0) if you want a timeout for getting data.
        epoch, data = await queue.get()
        if data is None:
            logger.info(
                "Got message from client about disconnection. Exiting consumer loop..."
            )
            break
        else:
            logger.info(f"Received callback data via async queue at {epoch}: {data}")


async def main(address: str, char_uuid: str):
    queue = asyncio.Queue()
    client_task = run_ble_client(address, char_uuid, queue)
    consumer_task = run_queue_consumer(queue)
    await asyncio.gather(client_task, consumer_task)
    logger.info("Main method done.")


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    asyncio.run(
        main(
            sys.argv[1] if len(sys.argv) > 1 else ADDRESS,
            sys.argv[2] if len(sys.argv) > 2 else CHARACTERISTIC_UUID,
        )
    )
```

<img width="1772" alt="Screen Shot 2022-08-26 at 6 31 49 PM" src="https://user-images.githubusercontent.com/55897268/186998931-6449003c-b899-4833-8cb8-f75b1f41cc03.png">

**Comment by dlech** [comment]
Which operating system are you using?

**Reply by htcrane** [reply]
macOS Monterey v12.4

**Reply by dlech** [reply]
The CoreBluetooth APIs on macOS use the same callback for reads and notifications, so unfortunately, if you are using both for the same characteristic, there isn't a good way to know if the callback was for the read request or from a notification. You are welcome to dig into the Bleak source code to see if you can sort it out better than it is already. Or maybe an alternative would be to use two different characteristics, one to notify and one to read.
